File: backend/app/services/srs.py
from datetime import datetime, timedelta

from ..models import Flashcard


def schedule_review(card: Flashcard, rating: int, now: datetime | None = None) -> Flashcard:
    if now is None:
        now = datetime.utcnow()

    if rating < 3:
        card.repetition = 0
        card.interval_days = 1
    else:
        if card.repetition == 0:
            card.interval_days = 1
        elif card.repetition == 1:
            card.interval_days = 6
        else:
            card.interval_days = round(card.interval_days * card.ease_factor)
        card.repetition += 1

    delta = 0.1 - (5 - rating) * (0.08 + (5 - rating) * 0.02)
    card.ease_factor = max(1.3, card.ease_factor + delta)
    card.due_at = now + timedelta(days=card.interval_days)
    card.updated_at = now
    return card

File: frontend/src/utils/srs.ts
import type { Flashcard, ReviewRating } from "../types";

export function scheduleNext(
  card: Flashcard,
  rating: ReviewRating,
  now: Date = new Date()
): Flashcard {
  let repetition = card.repetition;
  let intervalDays = card.intervalDays;
  let easeFactor = card.easeFactor;

  if (rating < 3) {
    repetition = 0;
    intervalDays = 1;
  } else {
    if (repetition === 0) {
      intervalDays = 1;
    } else if (repetition === 1) {
      intervalDays = 6;
    } else {
      intervalDays = Math.round(intervalDays * easeFactor);
    }
    repetition += 1;
  }

  const delta = 0.1 - (5 - rating) * (0.08 + (5 - rating) * 0.02);
  easeFactor = Math.max(1.3, easeFactor + delta);

  const dueAt = new Date(now);
  dueAt.setDate(dueAt.getDate() + intervalDays);

  return {
    ...card,
    repetition,
    intervalDays,
    easeFactor,
    dueAt: dueAt.toISOString()
  };
}
